File: letTxt-2/rotas.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "rotas.h"

Rota* criarRota(int id, const char* nome, const char* tipo, const char* regiao) {
    Rota* nova = (Rota*)malloc(sizeof(Rota));
    if (nova == NULL) {
        printf("Erro ao alocar memoria\n");
        return NULL;
    }
    
    nova->id = id;
    strcpy(nova->nome, nome);
    strcpy(nova->tipo, tipo);
    strcpy(nova->regiao, regiao);
    nova->proximo = NULL;
    
    return nova;
}

void adicionarRota(Rota** lista, Rota* nova) {
    if (*lista == NULL) {
        *lista = nova;
        return;
    }
    
    Rota* atual = *lista;
    while (atual->proximo != NULL) {
        atual = atual->proximo;
    }
    atual->proximo = nova;
}

void carregarRotas(Rota** lista, const char* filtro) {
    FILE* arquivo = fopen("rotas_de_transportes.txt", "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir arquivo\n");
        return;
    }
    
    char linha[256];
    while (fgets(linha, sizeof(linha), arquivo)) {
        int id;
        char nome[100], tipo[20], regiao[50];
        
        sscanf(linha, "%d;%[^;];%[^;];%[^\n]", &id, nome, tipo, regiao);
        
        if (strstr(tipo, filtro) != NULL || strstr(regiao, filtro) != NULL) {
            Rota* nova = criarRota(id, nome, tipo, regiao);
            adicionarRota(lista, nova);
        }
    }
    
    fclose(arquivo);
}

void mostrarRotas(Rota* lista) {
    if (lista == NULL) {
        printf("Nenhuma rota encontrada\n");
        return;
    }
    
    Rota* atual = lista;
    while (atual != NULL) {
        printf("ID: %d | Nome: %s | Tipo: %s | Regiao: %s\n",
               atual->id, atual->nome, atual->tipo, atual->regiao);
        atual = atual->proximo;
    }
}

void liberarRotas(Rota** lista) {
    Rota* atual = *lista;
    while (atual != NULL) {
        Rota* proximo = atual->proximo;
        free(atual);
        atual = proximo;
    }
    *lista = NULL;
}

void adicionarNovaRota(Rota** lista) {
    int id;
    char nome[100], tipo[20], regiao[50];
    
    printf("\nDigite o ID: ");
    scanf("%d", &id);
    getchar();
    
    printf("Digite o nome: ");
    scanf("%[^\n]", nome);
    getchar();
    
    printf("Digite o tipo: ");
    scanf("%[^\n]", tipo);
    getchar();
    
    printf("Digite a regiao: ");
    scanf("%[^\n]", regiao);
    getchar();
    
    Rota* nova = criarRota(id, nome, tipo, regiao);
    adicionarRota(lista, nova);
    
    FILE* arquivo = fopen("rotas_de_transportes.txt", "a");
    if (arquivo != NULL) {
        fprintf(arquivo, "%d;%s;%s;%s\n", id, nome, tipo, regiao);
        fclose(arquivo);
        printf("Rota adicionada com sucesso!\n");
    }
} 

File: letTxt-2-Python/rotas.py
# letTxt-2-Python/rotas.py
import csv
import os

# Define o nome do arquivo de dados
NOME_ARQUIVO = "rotas_de_transportes.txt"

class Rota:
    """Representa uma rota de transporte."""
    def __init__(self, id: int, nome: str, tipo: str, regiao: str):
        self.id = id
        self.nome = nome
        self.tipo = tipo
        self.regiao = regiao

    def __str__(self):
        return f"ID: {self.id} | Nome: {self.nome} | Tipo: {self.tipo} | Regiao: {self.regiao}"

def carregar_rotas(filtro: str = "", arquivo_path: str = NOME_ARQUIVO) -> list[Rota]:
    """
    Carrega rotas do arquivo CSV, opcionalmente filtrando por tipo ou região.
    Retorna uma lista de objetos Rota.
    """
    rotas_carregadas = []
    try:
        # Garante que o diretório do script seja considerado se o path for relativo
        script_dir = os.path.dirname(__file__) if "__file__" in locals() else "."
        path_completo = os.path.join(script_dir, arquivo_path)

        # Cria o diretório se não existir (necessário se o script for rodado de fora)
        os.makedirs(os.path.dirname(path_completo), exist_ok=True)


        with open(path_completo, 'r', encoding='utf-8', newline='') as arquivo:
            leitor = csv.reader(arquivo, delimiter=';')
            for linha in leitor:
                if len(linha) == 4:
                    try:
                        id_rota = int(linha[0])
                        nome, tipo, regiao = linha[1], linha[2], linha[3]

                        # Aplica o filtro se ele não estiver vazio
                        if not filtro or filtro.lower() in tipo.lower() or filtro.lower() in regiao.lower():
                            rotas_carregadas.append(Rota(id_rota, nome, tipo, regiao))
                    except ValueError:
                        print(f"Aviso: Ignorando linha com ID inválido: {linha}")
                    except IndexError:
                         print(f"Aviso: Ignorando linha mal formatada: {linha}")

    except FileNotFoundError:
        # Se o arquivo não existe, retorna lista vazia (primeira execução talvez)
        print(f"Aviso: Arquivo '{path_completo}' não encontrado. Nenhuma rota carregada.")
    except Exception as e:
        print(f"Erro inesperado ao ler o arquivo: {e}")

    return rotas_carregadas

def mostrar_rotas(lista_rotas: list[Rota]):
    """Imprime as rotas formatadas na lista fornecida."""
    if not lista_rotas:
        print("Nenhuma rota encontrada para o filtro fornecido.")
        return

    print("\n--- Rotas Encontradas ---")
    for rota in lista_rotas:
        print(rota)
    print("------------------------")


def adicionar_nova_rota(arquivo_path: str = NOME_ARQUIVO):
    """
    Solicita informações do usuário para uma nova rota e a adiciona ao arquivo CSV.
    """
    print("\n--- Adicionar Nova Rota ---")
    while True:
        try:
            id_rota = int(input("Digite o ID: "))
            break
        except ValueError:
            print("ID inválido. Por favor, digite um número.")

    nome = input("Digite o nome: ")
    tipo = input("Digite o tipo: ")
    regiao = input("Digite a região: ")

    nova_rota = Rota(id_rota, nome, tipo, regiao)

    try:
        # Garante que o diretório do script seja considerado se o path for relativo
        script_dir = os.path.dirname(__file__) if "__file__" in locals() else "."
        path_completo = os.path.join(script_dir, arquivo_path)

        # Cria o diretório se não existir
        os.makedirs(os.path.dirname(path_completo), exist_ok=True)


        # Verifica se o arquivo existe para adicionar cabeçalho se necessário (embora o original não tivesse)
        # existe = os.path.exists(path_completo)

        with open(path_completo, 'a', encoding='utf-8', newline='') as arquivo:
            escritor = csv.writer(arquivo, delimiter=';')
            # O arquivo original não tem cabeçalho, então não adicionamos um aqui.
            # if not existe:
            #     escritor.writerow(["ID", "Nome", "Tipo", "Regiao"]) # Exemplo se quisesse cabeçalho
            escritor.writerow([nova_rota.id, nova_rota.nome, nova_rota.tipo, nova_rota.regiao])
        print("Rota adicionada com sucesso!")

    except IOError as e:
        print(f"Erro ao escrever no arquivo '{path_completo}': {e}")
    except Exception as e:
        print(f"Erro inesperado ao adicionar rota: {e}") 